src/fairseq2/utils/state.py

```diff
@@ -0,0 +1,156 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import (
+    Any,
+    Dict,
+    Mapping,
+    Optional,
+    Protocol,
+    Tuple,
+    final,
+    runtime_checkable,
+)
+
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.nn import Module
+from torch.optim import Optimizer
+
+from fairseq2.typing import finaloverride
+
+
+@runtime_checkable
+class Stateful(Protocol):
+    """Represents an object that follows the ``state_dict`` convention."""
+
+    def state_dict(self) -> Dict[str, Any]:
+        ...
+
+    def load_state_dict(self, state_dict: Mapping[str, Any]) -> None:
+        ...
+
+
+class StatefulObjectBag:
```

Comment:
Is this object intended to be used as a singleton ?

Reply:
No. Let me share the training recipe tomorrow. There you can see how it is used.